File: native/hub/src/app/entry/upload/upload_view_model.rs
use http::StatusCode;
use crate::bridge::{RustOperation, RustRequest, RustResponse, RustSignal};
use crate::messages::entry::upload::uploaded_content;
use prost::Message;
use tokio_with_wasm::tokio;
use crate::app::entry::upload::upload_model::{UploadModel, TextFile};




// Handler functions
pub async fn handle_uploaded_content(rust_request: RustRequest,
    upload_model: &mut tokio::sync::MutexGuard<'_, UploadModel>) -> RustResponse {
    use crate::messages::entry::upload::uploaded_content::{CreateRequest, CreateResponse, ReadRequest, ReadResponse};

    match rust_request.operation {
        RustOperation::Create => {
            let message_bytes = rust_request.message.unwrap();
            let request_message = CreateRequest::decode(message_bytes.as_slice()).unwrap();

            // Do something with data
            upload_model.file_paths = request_message.file_paths;
            upload_model.urls = request_message.urls;
            upload_model.text_files = request_message.texts
                .into_iter()
                .map(|prost_text_file| crate::app::entry::upload::upload_model::TextFile {
                    title: prost_text_file.title,
                    content: prost_text_file.content,
                })
                .collect();

            let response_message;
            if upload_model.file_paths.len() > 0 {
                response_message = CreateResponse {
                    // Send the data back in a response
                    status_code: StatusCode::OK.as_u16() as u32
                };
            }
            else {
                response_message = CreateResponse {
                    // Send the data back in a response
                    status_code: StatusCode::NOT_FOUND.as_u16() as u32
                };
            }
            
            RustResponse {
                successful: true,
                message: Some(response_message.encode_to_vec()),
                blob: None,
            }
        },
        RustOperation::Read => {
            // Debug purposes. Just to check if the uploaded files are stored in main().
            let message_bytes = rust_request.message.unwrap();
            let request_message = ReadRequest::decode(message_bytes.as_slice()).unwrap();

            let _ = request_message;

            let response_message = ReadResponse {
                file_paths: upload_model.file_paths.clone(),
                urls: upload_model.urls.clone(),
                texts: upload_model.text_files.clone()
                    .into_iter()
                    .map(|prost_text_file| uploaded_content::TextFile {
                        title: prost_text_file.title,
                        content: prost_text_file.content,
                    })
                    .collect()
            };

            RustResponse {
                successful: true,
                message: Some(response_message.encode_to_vec()),
                blob: None,
            }
        },
        RustOperation::Update => RustResponse::default(),
        RustOperation::Delete => RustResponse::default(),
    }
}

File: lib/src/lessonlab_modules/entry/upload/upload_view_model.dart
// ignore_for_file: unnecessary_getters_setters

import 'package:flutter/material.dart';
import 'package:lessonlab/src/lessonlab_modules/entry/menu/menu_view.dart';
import 'package:lessonlab/src/lessonlab_modules/lesson/lesson_specifications_view.dart';
import 'package:lessonlab/src/lessonlab_modules/quiz/quiz_specifications_view.dart';
import 'package:lessonlab/src/lessonlab_modules/results/lesson_result/lesson_result_view.dart';
import 'package:lessonlab/src/lessonlab_modules/entry/upload/components/overlay/overlay_controller.dart';
import 'package:cross_file/cross_file.dart';
import 'dart:developer' as developer;

import 'package:lessonlab/messages/entry/upload/uploaded_content.pb.dart'
    // ignore: library_prefixes
    as RinfInterface;
import 'package:rinf/rinf.dart';

class UploadViewModel with ChangeNotifier {
  final _files = <XFile>[];
  List<XFile> get files => _files;

  final _urls = <String>[];
  List<String> get urls => _urls;

  final _textFiles = <TextFile>[];
  List<TextFile> get textFiles => _textFiles;

  var _statusCode = 0;

  bool _hasFiles = false;
  bool get hasFiles => _hasFiles;
  set hasFiles(bool value) {
    _hasFiles = value;
  }

  Future<void> sendData() async {
    final requestMessage = RinfInterface.CreateRequest(
      filePaths: _files.map((file) => file.path).toList(),
      urls: _urls,
      texts: _textFiles
          .map((textFile) => RinfInterface.TextFile(
                title: textFile.title,
                content: textFile.content,
              ))
          .toList(),
    );
    final rustRequest = RustRequest(
      resource: RinfInterface.ID,
      operation: RustOperation.Create,
      message: requestMessage.writeToBuffer(),
      // blob: NO BLOB
    );
    final rustResponse = await requestToRust(rustRequest);
    final responseMessage = RinfInterface.CreateResponse.fromBuffer(
      rustResponse.message!,
    );
    _statusCode = responseMessage.statusCode;
    developer.log(_statusCode.toString(), name: 'rinf-info');
  }

  Future<void> getData() async {
    // Debug purposes. Just to check if the uploaded files are stored in rust main().
    final requestMessage = RinfInterface.ReadRequest(req: true);
    final rustRequest = RustRequest(
      resource: RinfInterface.ID,
      operation: RustOperation.Read,
      message: requestMessage.writeToBuffer(),
      // blob: NO BLOB
    );
    final rustResponse = await requestToRust(rustRequest);
    final responseMessage = RinfInterface.ReadResponse.fromBuffer(
      rustResponse.message!,
    );
    var content = responseMessage.filePaths;
    developer.log(content.join(), name: 'rinf-info');
  }

  void cancelUpload(BuildContext context) {
    Navigator.restorablePushNamed(
      context,
      MenuView.routeName,
    );
  }

  void newLesson(BuildContext context) {
    if (_hasFiles) {
      sendData();
      getData();
      Navigator.restorablePushNamed(
        context,
        LessonSpecificationsView.routeName,
      );
    } else {
      null;
    }
  }

  void newQuiz(BuildContext context) {
    if (_hasFiles) {
      sendData();
      getData();
      Navigator.restorablePushNamed(
        context,
        QuizSpecificationsView.routeName,
      );
    } else {
      null;
    }
  }
}
